test/conformance/generatename_test.go

```diff
@@ -0,0 +1,187 @@
+// and serve requests.
+func TestServiceGenerateName(t *testing.T) {
+	t.Parallel()
+	clients := setup(t)
+
+	generateName := getGenerateNamePrefix(t)
+	names := test.ResourceNames{
+		Image: helloworld,
+	}
+
+	// Cleanup on test failure.
+	test.CleanupOnInterrupt(func() { test.TearDown(clients, names) })
+	defer func() { test.TearDown(clients, names) }()
+
+	// Create the service using the generate name field. If the serivce does not become ready this will fail.
+	t.Logf("Creating new service with generateName %s", generateName)
+	resources, err := test.CreateRunLatestServiceReady(t, clients, &names, &test.Options{}, setServiceGenerateName(generateName))
+	if err != nil {
+		t.Fatalf("Failed to create service with generateName %s: %v", generateName, err)
+	}
+
+	// Ensure that the name given to the service is generated from the generateName field.
+	t.Log("When the service is created, the name is generated using the provided generateName")
+	if names.Service == generateName {
+		t.Fatalf("Service did not recieve a randomized name, instead got %s", names.Service)
+	}
+	if !strings.HasPrefix(names.Service, generateName) {
+		t.Fatalf("Service name is not prefixed by the generateName field, expected %s to be a prefix, got %s", generateName, names.Service)
+	}
+
+	// Ensure that the service can serve requests
+	ensureCanServeRequests(t, clients, resources.Route)
+}
+
+// TestRouteAndConfiguration checks that both routes and configurations MAY request names generated by
+// the system using metadata.generateName. It ensures that routes and configurations created this way both:
+// 1. Become ready
+// 2. Can serve requests.
+func TestRouteAndConfigurationGenerateName(t *testing.T) {
+	t.Parallel()
+	clients := setup(t)
+
+	generateName := getGenerateNamePrefix(t)
+	names := test.ResourceNames{
+		Image: helloworld,
+	}
+
+	test.CleanupOnInterrupt(func() { test.TearDown(clients, names) })
+	defer func() { test.TearDown(clients, names) }()
+
+	t.Logf("Creating new configuration with generateName %s", generateName)
+	config, err := test.CreateConfiguration(t, clients, names, &test.Options{}, setConfigurationGenerateName(generateName))
+	if err != nil {
+		t.Fatalf("Failed to create configuration with generateName %s: %v", generateName, err)
+	}
+	names.Config = config.Name
+
+	// Ensure the associated revision is created. This also checks that the configuration becomes ready.
+	t.Log("The configuration will be updated with the name of the associated Revision once it is created.")
+	names.Revision, err = test.WaitForConfigLatestRevision(clients, names)
+	if err != nil {
+		t.Fatalf("Configuration %s was not updated with the new revision: %v", names.Config, err)
+	}
+
+	// Ensure that the name given to the configuration is generated from the generate name field.
+	t.Log("When the configuration is created, the name is generated using the provided generateName")
+	if names.Config == generateName {
```

Comment:
These statements could probably be refactored out to a common function for Service, Config, and Route to reduce duplication (i.e. `validateName(prefix, name string) error` . A regex also might be both simpler and capture more requirements as to what we expect from generate name such as:

1. What characters should follow the prefix?
2. How long should the generated name be?

Reply:
Moved to `validateName`, and changed it to a regex. It now validates that the prefix is followed by 5 lower-case characters or digits, which is what the current behaviour is.

It might be a bit overly strict since it now depends on the api-server implementation not changing. I believe Kubernetes only guarantees a unique suffix (https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#idempotency). Is this updated regex okay?

Reply:
I would loosen it up a bit to be prefix followed by one or more valid name characters ( https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names )

Reply:
Okay, loosed up the regex and changed it. Thanks.